File: mobile/src/main/java/com/alexstyl/specialdates/addevent/ui/ContactsSearch.java
package com.alexstyl.specialdates.addevent.ui;

import com.alexstyl.specialdates.DisplayName;
import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.search.NameMatcher;
import com.alexstyl.specialdates.service.PeopleEventsProvider;
import com.alexstyl.specialdates.upcoming.TimePeriod;

import java.util.ArrayList;
import java.util.List;

final class ContactsSearch {

    private final PeopleEventsProvider peopleEventsProvider;
    private final NameMatcher nameMatcher;

    ContactsSearch(PeopleEventsProvider peopleEventsProvider, NameMatcher nameMatcher) {
        this.peopleEventsProvider = peopleEventsProvider;
        this.nameMatcher = nameMatcher;
    }

    List<Contact> searchForContacts(String searchQuery, int counter) {
        searchQuery = searchQuery.trim();
        List<Contact> matchedContacts = new ArrayList<>();
        Date from = Date.today();
        Date to = from.addDay(364);
        List<ContactEvent> contactEvents = peopleEventsProvider.getCelebrationDateFor(TimePeriod.between(from, to));
        int size = 0;

        for (ContactEvent contactEvent : contactEvents) {
            DisplayName displayName = contactEvent.getContact().getDisplayName();
            if (nameMatcher.match(displayName, searchQuery)) {
                matchedContacts.add(contactEvent.getContact());
                size++;
                if (size == counter) {
                    break;
                }
            }
        }
        return matchedContacts;
    }

}

File: memento/src/main/java/com/alexstyl/specialdates/addevent/ContactsSearch.kt
package com.alexstyl.specialdates.addevent

import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.contact.ContactsProvider
import com.alexstyl.specialdates.search.NameMatcher
import java.util.ArrayList

internal class ContactsSearch(private val contactsProvider: ContactsProvider, private val nameMatcher: NameMatcher) {

    fun searchForContacts(searchQuery: String, numberOfResults: Int): List<Contact> {
        var searchQuery = searchQuery
        val matchedContacts = ArrayList<Contact>()
        if (numberOfResults == 0) {
            return matchedContacts
        }
        searchQuery = searchQuery.trim { it <= ' ' }
        val allContacts = contactsProvider.allContacts
        if (allContacts.isEmpty()) {
            return matchedContacts
        }
        var count = 0
        var index = 0
        val contactSize = allContacts.size
        do {
            val contact = allContacts[index]
            val displayName = contact.displayName
            if (nameMatcher.match(displayName, searchQuery)) {
                matchedContacts.add(contact)
                count++
            }
            index++
        } while (count < numberOfResults && index < contactSize)

        return matchedContacts
    }

}
